metrics/isolator.cpp

```diff
@@ -0,0 +1,580 @@
+  {
+    // Set `serviceScheme` based on flags.
+    ASSERT(flags.service_scheme.isSome());
+    ASSERT(flags.service_scheme.get() == "http" ||
+           flags.service_scheme.get() == "https");
+    serviceScheme = flags.service_scheme.get();
+
+    // Set `service*Address` based on flags.
+    ASSERT(flags.service_address.isSome());
+
+    ASSERT(flags.service_network.isSome());
+    ASSERT(flags.service_network.get() == "inet" ||
+           flags.service_network.get() == "unix");
+
+    if (flags.service_network.get() == "inet") {
+      vector<string> hostport =
+        strings::split(flags.service_address.get(), ":");
+      if (hostport.size() != 2) {
+        LOG(FATAL) << "Unable to split '" << flags.service_address.get() << "'"
+                   << " into valid 'host:port' combination";
+      }
+
+      Try<net::IP> ip = net::IP::parse(hostport[0]);
+      if (ip.isError()) {
+        LOG(FATAL) << "Unable to parse '" << hostport[0] << "'"
+                   << " as a valid IP address: " << ip.error();
+      }
+
+      Try<uint16_t> port = numify<uint16_t>(hostport[1]);
+      if (port.isError()) {
+        LOG(FATAL) << "Unable parse '" + hostport[1] + "'"
+                   << " as a valid port of type 'uint16_t': " + port.error();
+      }
+
+      serviceInetAddress = inet::Address(ip.get(), port.get());
+    }
+
+    if (flags.service_network.get() == "unix") {
+      Try<unix::Address> address =
+        unix::Address::create(flags.service_address.get());
+      if (address.isError()) {
+        LOG(FATAL) << "Unable to convert '" + flags.service_address.get() + "'"
+                   << " to valid 'unix' address: " + address.error();
+      }
+
+      serviceUnixAddress = address.get();
+    }
+
+    // Set `serviceEndpoint` based on flags.
+    ASSERT(flags.service_endpoint.isSome());
+    serviceEndpoint = flags.service_endpoint.get();
+  }
+
+  // Search through `legacyStateDir` and let the DC/OS metrics service
+  // know about any recovered containers that still have state in there.
+  virtual Future<Nothing> recover(
+      const vector<ContainerState>& states,
+      const hashset<ContainerID>& orphans)
+  {
+    // If there is no `legacyStateDir`, we are done.
+    if (legacyStateDir.isNone()) {
+      return Nothing();
+    }
+
+    // If the legacy state directory has been set in the configuration but the
+    // directory is not found, then we assume recovery was previously completed
+    // successfully. Return successfully in this case.
+    if (!os::exists(legacyStateDir.get())) {
+      return Nothing();
+    }
+
+    // Otherwise collect all containers known by either the agent
+    // (states) or the containerizer (orphans).
+    vector<ContainerID> containers;
+
+    foreach(const ContainerState& state, states) {
+      containers.push_back(state.container_id());
+    }
+
+    foreach(const ContainerID& container, orphans) {
+      containers.push_back(container);
+    }
```

Comment:
Note: It probably isn't a good idea to return a failure if the `cleanup` method fails, because that would abort the other isolators' cleanups, including important ones like the cgroups isolator.